File: Kernel Simulator/Misc/Screensaver/Displays/FallingLine.vb

'    Kernel Simulator  Copyright (C) 2018-2022  Aptivi
'
'    This file is part of Kernel Simulator
'
'    Kernel Simulator is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    Kernel Simulator is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Namespace Misc.Screensaver.Displays
    Public Module FallingLineSettings

        Private _fallingLine255Colors As Boolean
        Private _fallingLineTrueColor As Boolean = True
        Private _fallingLineDelay As Integer = 10
        Private _fallingLineMaxSteps As Integer = 25
        Private _fallingLineMinimumRedColorLevel As Integer = 0
        Private _fallingLineMinimumGreenColorLevel As Integer = 0
        Private _fallingLineMinimumBlueColorLevel As Integer = 0
        Private _fallingLineMinimumColorLevel As Integer = 0
        Private _fallingLineMaximumRedColorLevel As Integer = 255
        Private _fallingLineMaximumGreenColorLevel As Integer = 255
        Private _fallingLineMaximumBlueColorLevel As Integer = 255
        Private _fallingLineMaximumColorLevel As Integer = 255

        ''' <summary>
        ''' [FallingLine] Enable 255 color support. Has a higher priority than 16 color support.
        ''' </summary>
        Public Property FallingLine255Colors As Boolean
            Get
                Return _fallingLine255Colors
            End Get
            Set(value As Boolean)
                _fallingLine255Colors = value
            End Set
        End Property
        ''' <summary>
        ''' [FallingLine] Enable truecolor support. Has a higher priority than 255 color support.
        ''' </summary>
        Public Property FallingLineTrueColor As Boolean
            Get
                Return _fallingLineTrueColor
            End Get
            Set(value As Boolean)
                _fallingLineTrueColor = value
            End Set
        End Property
        ''' <summary>
        ''' [FallingLine] How many milliseconds to wait before making the next write?
        ''' </summary>
        Public Property FallingLineDelay As Integer
            Get
                Return _fallingLineDelay
            End Get
            Set(value As Integer)
                If value <= 0 Then value = 10
                _fallingLineDelay = value
            End Set
        End Property
        ''' <summary>
        ''' [FallingLine] How many fade steps to do?
        ''' </summary>
        Public Property FallingLineMaxSteps As Integer
            Get
                Return _fallingLineMaxSteps
            End Get
            Set(value As Integer)
                If value <= 0 Then value = 25
                _fallingLineMaxSteps = value
            End Set
        End Property
        ''' <summary>
        ''' [FallingLine] The minimum red color level (true color)
        ''' </summary>
        Public Property FallingLineMinimumRedColorLevel As Integer
            Get
                Return _fallingLineMinimumRedColorLevel
            End Get
            Set(value As Integer)
                If value <= 0 Then value = 0
                If value > 255 Then value = 255
                _fallingLineMinimumRedColorLevel = value
            End Set
        End Property
        ''' <summary>
        ''' [FallingLine] The minimum green color level (true color)
        ''' </summary>
        Public Property FallingLineMinimumGreenColorLevel As Integer
            Get
                Return _fallingLineMinimumGreenColorLevel
            End Get
            Set(value As Integer)
                If value <= 0 Then value = 0
                If value > 255 Then value = 255
                _fallingLineMinimumGreenColorLevel = value
            End Set
        End Property
        ''' <summary>
        ''' [FallingLine] The minimum blue color level (true color)
        ''' </summary>
        Public Property FallingLineMinimumBlueColorLevel As Integer
            Get
                Return _fallingLineMinimumBlueColorLevel
            End Get
            Set(value As Integer)
                If value <= 0 Then value = 0
                If value > 255 Then value = 255
                _fallingLineMinimumBlueColorLevel = value
            End Set
        End Property
        ''' <summary>
        ''' [FallingLine] The minimum color level (255 colors or 16 colors)
        ''' </summary>
        Public Property FallingLineMinimumColorLevel As Integer
            Get
                Return _fallingLineMinimumColorLevel
            End Get
            Set(value As Integer)
                Dim FinalMinimumLevel As Integer = If(_fallingLine255Colors Or _fallingLineTrueColor, 255, 15)
                If value <= 0 Then value = 0
                If value > FinalMinimumLevel Then value = FinalMinimumLevel
                _fallingLineMinimumColorLevel = value
            End Set
        End Property
        ''' <summary>
        ''' [FallingLine] The maximum red color level (true color)
        ''' </summary>
        Public Property FallingLineMaximumRedColorLevel As Integer
            Get
                Return _fallingLineMaximumRedColorLevel
            End Get
            Set(value As Integer)
                If value <= _fallingLineMinimumRedColorLevel Then value = _fallingLineMinimumRedColorLevel
                If value > 255 Then value = 255
                _fallingLineMaximumRedColorLevel = value
            End Set
        End Property
        ''' <summary>
        ''' [FallingLine] The maximum green color level (true color)
        ''' </summary>
        Public Property FallingLineMaximumGreenColorLevel As Integer
            Get
                Return _fallingLineMaximumGreenColorLevel
            End Get
            Set(value As Integer)
                If value <= _fallingLineMinimumGreenColorLevel Then value = _fallingLineMinimumGreenColorLevel
                If value > 255 Then value = 255
                _fallingLineMaximumGreenColorLevel = value
            End Set
        End Property
        ''' <summary>
        ''' [FallingLine] The maximum blue color level (true color)
        ''' </summary>
        Public Property FallingLineMaximumBlueColorLevel As Integer
            Get
                Return _fallingLineMaximumBlueColorLevel
            End Get
            Set(value As Integer)
                If value <= _fallingLineMinimumBlueColorLevel Then value = _fallingLineMinimumBlueColorLevel
                If value > 255 Then value = 255
                _fallingLineMaximumBlueColorLevel = value
            End Set
        End Property
        ''' <summary>
        ''' [FallingLine] The maximum color level (255 colors or 16 colors)
        ''' </summary>
        Public Property FallingLineMaximumColorLevel As Integer
            Get
                Return _fallingLineMaximumColorLevel
            End Get
            Set(value As Integer)
                Dim FinalMaximumLevel As Integer = If(_fallingLine255Colors Or _fallingLineTrueColor, 255, 15)
                If value <= _fallingLineMinimumColorLevel Then value = _fallingLineMinimumColorLevel
                If value > FinalMaximumLevel Then value = FinalMaximumLevel
                _fallingLineMaximumColorLevel = value
            End Set
        End Property

    End Module

    Public Class FallingLineDisplay
        Inherits BaseScreensaver
        Implements IScreensaver

        Private RandomDriver As Random
        Private ColumnLine As Integer
        Private CurrentWindowWidth As Integer
        Private CurrentWindowHeight As Integer
        Private ResizeSyncing As Boolean
        Private ReadOnly CoveredPositions As New List(Of Tuple(Of Integer, Integer))

        Public Overrides Property ScreensaverName As String = "FallingLine" Implements IScreensaver.ScreensaverName

        Public Overrides Property ScreensaverSettings As Dictionary(Of String, Object) Implements IScreensaver.ScreensaverSettings

        Public Overrides Sub ScreensaverPreparation() Implements IScreensaver.ScreensaverPreparation
            'Variable preparations
            RandomDriver = New Random
            CurrentWindowWidth = Console.WindowWidth
            CurrentWindowHeight = Console.WindowHeight
            Console.BackgroundColor = ConsoleColor.Black
            Console.ForegroundColor = ConsoleColor.White
            Console.Clear()
            Console.CursorVisible = False
        End Sub

        Public Overrides Sub ScreensaverLogic() Implements IScreensaver.ScreensaverLogic
            'Choose the column for the falling line
            ColumnLine = RandomDriver.Next(Console.WindowWidth)

            'Now, determine the fall start and end position
            Dim FallStart As Integer = 0
            Dim FallEnd As Integer = Console.WindowHeight - 1

            'Select the color
            Dim ColorStorage As Color
            If FallingLineTrueColor Then
                Dim RedColorNum As Integer = RandomDriver.Next(FallingLineMinimumRedColorLevel, FallingLineMaximumRedColorLevel)
                Dim GreenColorNum As Integer = RandomDriver.Next(FallingLineMinimumGreenColorLevel, FallingLineMaximumGreenColorLevel)
                Dim BlueColorNum As Integer = RandomDriver.Next(FallingLineMinimumBlueColorLevel, FallingLineMaximumBlueColorLevel)
                WdbgConditional(ScreensaverDebug, DebugLevel.I, "Got color (R;G;B: {0};{1};{2})", RedColorNum, GreenColorNum, BlueColorNum)
                ColorStorage = New Color(RedColorNum, GreenColorNum, BlueColorNum)
                SetConsoleColor(ColorStorage, True)
            ElseIf FallingLine255Colors Then
                Dim ColorNum As Integer = RandomDriver.Next(FallingLineMinimumColorLevel, FallingLineMaximumColorLevel)
                WdbgConditional(ScreensaverDebug, DebugLevel.I, "Got color ({0})", ColorNum)
                ColorStorage = New Color(ColorNum)
                SetConsoleColor(ColorStorage, True)
            Else
                Console.BackgroundColor = colors(RandomDriver.Next(FallingLineMinimumColorLevel, FallingLineMaximumColorLevel))
                WdbgConditional(ScreensaverDebug, DebugLevel.I, "Got color ({0})", Console.BackgroundColor)
                ColorStorage = New Color(Console.BackgroundColor)
            End If

            'Make the line fall down
            For Fall As Integer = FallStart To FallEnd
                'Check to see if user decided to resize
                If CurrentWindowHeight <> Console.WindowHeight Or CurrentWindowWidth <> Console.WindowWidth Then ResizeSyncing = True
                If ResizeSyncing Then Exit For

                'Print a block and add the covered position to the list so fading down can be done
                WriteWherePlain(" ", ColumnLine, Fall, False)
                Dim PositionTuple As New Tuple(Of Integer, Integer)(ColumnLine, Fall)
                CoveredPositions.Add(PositionTuple)

                'Delay
                SleepNoBlock(FallingLineDelay, ScreensaverDisplayerThread)
            Next

            'Fade the line down. Please note that this requires true-color support in the terminal to work properly.
            For StepNum As Integer = 0 To FallingLineMaxSteps
                'Check to see if user decided to resize
                If CurrentWindowHeight <> Console.WindowHeight Or CurrentWindowWidth <> Console.WindowWidth Then ResizeSyncing = True
                If ResizeSyncing Then Exit For

                'Set thresholds
                Dim ThresholdRed As Double = ColorStorage.R / FallingLineMaxSteps
                Dim ThresholdGreen As Double = ColorStorage.G / FallingLineMaxSteps
                Dim ThresholdBlue As Double = ColorStorage.B / FallingLineMaxSteps
                WdbgConditional(ScreensaverDebug, DebugLevel.I, "Color threshold (R;G;B: {0})", ThresholdRed, ThresholdGreen, ThresholdBlue)

                'Set color fade steps
                Dim CurrentColorRedOut As Integer = ColorStorage.R - ThresholdRed * StepNum
                Dim CurrentColorGreenOut As Integer = ColorStorage.G - ThresholdGreen * StepNum
                Dim CurrentColorBlueOut As Integer = ColorStorage.B - ThresholdBlue * StepNum
                WdbgConditional(ScreensaverDebug, DebugLevel.I, "Color out (R;G;B: {0};{1};{2})", CurrentColorRedOut, CurrentColorGreenOut, CurrentColorBlueOut)

                'Get the positions and write the block with new color
                Dim CurrentFadeColor As New Color(CurrentColorRedOut, CurrentColorGreenOut, CurrentColorBlueOut)
                For Each PositionTuple As Tuple(Of Integer, Integer) In CoveredPositions
                    'Check to see if user decided to resize
                    If CurrentWindowHeight <> Console.WindowHeight Or CurrentWindowWidth <> Console.WindowWidth Then ResizeSyncing = True
                    If ResizeSyncing Then Exit For

                    'Actually fade the line out
                    Dim PositionLeft As Integer = PositionTuple.Item1
                    Dim PositionTop As Integer = PositionTuple.Item2
                    WriteWhere(" ", PositionLeft, PositionTop, False, Color.Empty, CurrentFadeColor)
                Next

                'Delay
                SleepNoBlock(FallingLineDelay, ScreensaverDisplayerThread)
            Next

            'Reset resize sync
            CoveredPositions.Clear()
            ResizeSyncing = False
            CurrentWindowWidth = Console.WindowWidth
            CurrentWindowHeight = Console.WindowHeight
            SleepNoBlock(FallingLineDelay, ScreensaverDisplayerThread)
        End Sub

    End Class
End Namespace

File: Kernel Simulator/Misc/Screensaver/Displays/FallingLine.cs
using System;
using System.Collections.Generic;
using KS.ConsoleBase.Colors;
using KS.Misc.Threading;
using KS.Misc.Writers.ConsoleWriters;
using KS.Misc.Writers.DebugWriters;
using Terminaux.Base;
using Terminaux.Colors;

// Kernel Simulator  Copyright (C) 2018-2022  Aptivi
// 
// This file is part of Kernel Simulator
// 
// Kernel Simulator is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// Kernel Simulator is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

namespace KS.Misc.Screensaver.Displays
{
	public static class FallingLineSettings
	{

		private static bool _fallingLine255Colors;
		private static bool _fallingLineTrueColor = true;
		private static int _fallingLineDelay = 10;
		private static int _fallingLineMaxSteps = 25;
		private static int _fallingLineMinimumRedColorLevel = 0;
		private static int _fallingLineMinimumGreenColorLevel = 0;
		private static int _fallingLineMinimumBlueColorLevel = 0;
		private static int _fallingLineMinimumColorLevel = 0;
		private static int _fallingLineMaximumRedColorLevel = 255;
		private static int _fallingLineMaximumGreenColorLevel = 255;
		private static int _fallingLineMaximumBlueColorLevel = 255;
		private static int _fallingLineMaximumColorLevel = 255;

		/// <summary>
		/// [FallingLine] Enable 255 color support. Has a higher priority than 16 color support.
		/// </summary>
		public static bool FallingLine255Colors
		{
			get
			{
				return _fallingLine255Colors;
			}
			set
			{
				_fallingLine255Colors = value;
			}
		}
		/// <summary>
		/// [FallingLine] Enable truecolor support. Has a higher priority than 255 color support.
		/// </summary>
		public static bool FallingLineTrueColor
		{
			get
			{
				return _fallingLineTrueColor;
			}
			set
			{
				_fallingLineTrueColor = value;
			}
		}
		/// <summary>
		/// [FallingLine] How many milliseconds to wait before making the next write?
		/// </summary>
		public static int FallingLineDelay
		{
			get
			{
				return _fallingLineDelay;
			}
			set
			{
				if (value <= 0)
					value = 10;
				_fallingLineDelay = value;
			}
		}
		/// <summary>
		/// [FallingLine] How many fade steps to do?
		/// </summary>
		public static int FallingLineMaxSteps
		{
			get
			{
				return _fallingLineMaxSteps;
			}
			set
			{
				if (value <= 0)
					value = 25;
				_fallingLineMaxSteps = value;
			}
		}
		/// <summary>
		/// [FallingLine] The minimum red color level (true color)
		/// </summary>
		public static int FallingLineMinimumRedColorLevel
		{
			get
			{
				return _fallingLineMinimumRedColorLevel;
			}
			set
			{
				if (value <= 0)
					value = 0;
				if (value > 255)
					value = 255;
				_fallingLineMinimumRedColorLevel = value;
			}
		}
		/// <summary>
		/// [FallingLine] The minimum green color level (true color)
		/// </summary>
		public static int FallingLineMinimumGreenColorLevel
		{
			get
			{
				return _fallingLineMinimumGreenColorLevel;
			}
			set
			{
				if (value <= 0)
					value = 0;
				if (value > 255)
					value = 255;
				_fallingLineMinimumGreenColorLevel = value;
			}
		}
		/// <summary>
		/// [FallingLine] The minimum blue color level (true color)
		/// </summary>
		public static int FallingLineMinimumBlueColorLevel
		{
			get
			{
				return _fallingLineMinimumBlueColorLevel;
			}
			set
			{
				if (value <= 0)
					value = 0;
				if (value > 255)
					value = 255;
				_fallingLineMinimumBlueColorLevel = value;
			}
		}
		/// <summary>
		/// [FallingLine] The minimum color level (255 colors or 16 colors)
		/// </summary>
		public static int FallingLineMinimumColorLevel
		{
			get
			{
				return _fallingLineMinimumColorLevel;
			}
			set
			{
				int FinalMinimumLevel = _fallingLine255Colors | _fallingLineTrueColor ? 255 : 15;
				if (value <= 0)
					value = 0;
				if (value > FinalMinimumLevel)
					value = FinalMinimumLevel;
				_fallingLineMinimumColorLevel = value;
			}
		}
		/// <summary>
		/// [FallingLine] The maximum red color level (true color)
		/// </summary>
		public static int FallingLineMaximumRedColorLevel
		{
			get
			{
				return _fallingLineMaximumRedColorLevel;
			}
			set
			{
				if (value <= _fallingLineMinimumRedColorLevel)
					value = _fallingLineMinimumRedColorLevel;
				if (value > 255)
					value = 255;
				_fallingLineMaximumRedColorLevel = value;
			}
		}
		/// <summary>
		/// [FallingLine] The maximum green color level (true color)
		/// </summary>
		public static int FallingLineMaximumGreenColorLevel
		{
			get
			{
				return _fallingLineMaximumGreenColorLevel;
			}
			set
			{
				if (value <= _fallingLineMinimumGreenColorLevel)
					value = _fallingLineMinimumGreenColorLevel;
				if (value > 255)
					value = 255;
				_fallingLineMaximumGreenColorLevel = value;
			}
		}
		/// <summary>
		/// [FallingLine] The maximum blue color level (true color)
		/// </summary>
		public static int FallingLineMaximumBlueColorLevel
		{
			get
			{
				return _fallingLineMaximumBlueColorLevel;
			}
			set
			{
				if (value <= _fallingLineMinimumBlueColorLevel)
					value = _fallingLineMinimumBlueColorLevel;
				if (value > 255)
					value = 255;
				_fallingLineMaximumBlueColorLevel = value;
			}
		}
		/// <summary>
		/// [FallingLine] The maximum color level (255 colors or 16 colors)
		/// </summary>
		public static int FallingLineMaximumColorLevel
		{
			get
			{
				return _fallingLineMaximumColorLevel;
			}
			set
			{
				int FinalMaximumLevel = _fallingLine255Colors | _fallingLineTrueColor ? 255 : 15;
				if (value <= _fallingLineMinimumColorLevel)
					value = _fallingLineMinimumColorLevel;
				if (value > FinalMaximumLevel)
					value = FinalMaximumLevel;
				_fallingLineMaximumColorLevel = value;
			}
		}

	}

	public class FallingLineDisplay : BaseScreensaver, IScreensaver
	{

		private Random RandomDriver;
		private int ColumnLine;
		private int CurrentWindowWidth;
		private int CurrentWindowHeight;
		private bool ResizeSyncing;
		private readonly List<Tuple<int, int>> CoveredPositions = [];

		public override string ScreensaverName { get; set; } = "FallingLine";

		public override Dictionary<string, object> ScreensaverSettings { get; set; }

		public override void ScreensaverPreparation()
		{
			// Variable preparations
			RandomDriver = new Random();
			CurrentWindowWidth = ConsoleWrapper.WindowWidth;
			CurrentWindowHeight = ConsoleWrapper.WindowHeight;
			Console.BackgroundColor = ConsoleColor.Black;
			Console.ForegroundColor = ConsoleColor.White;
			ConsoleWrapper.Clear();
			ConsoleWrapper.CursorVisible = false;
		}

		public override void ScreensaverLogic()
		{
			// Choose the column for the falling line
			ColumnLine = RandomDriver.Next(ConsoleWrapper.WindowWidth);

			// Now, determine the fall start and end position
			int FallStart = 0;
			int FallEnd = ConsoleWrapper.WindowHeight - 1;

			// Select the color
			Color ColorStorage;
			if (FallingLineSettings.FallingLineTrueColor)
			{
				int RedColorNum = RandomDriver.Next(FallingLineSettings.FallingLineMinimumRedColorLevel, FallingLineSettings.FallingLineMaximumRedColorLevel);
				int GreenColorNum = RandomDriver.Next(FallingLineSettings.FallingLineMinimumGreenColorLevel, FallingLineSettings.FallingLineMaximumGreenColorLevel);
				int BlueColorNum = RandomDriver.Next(FallingLineSettings.FallingLineMinimumBlueColorLevel, FallingLineSettings.FallingLineMaximumBlueColorLevel);
				DebugWriter.WdbgConditional(ref Screensaver.ScreensaverDebug, DebugLevel.I, "Got color (R;G;B: {0};{1};{2})", RedColorNum, GreenColorNum, BlueColorNum);
				ColorStorage = new Color(RedColorNum, GreenColorNum, BlueColorNum);
				KernelColorTools.SetConsoleColor(ColorStorage, true);
			}
			else if (FallingLineSettings.FallingLine255Colors)
			{
				int ColorNum = RandomDriver.Next(FallingLineSettings.FallingLineMinimumColorLevel, FallingLineSettings.FallingLineMaximumColorLevel);
				DebugWriter.WdbgConditional(ref Screensaver.ScreensaverDebug, DebugLevel.I, "Got color ({0})", ColorNum);
				ColorStorage = new Color(ColorNum);
				KernelColorTools.SetConsoleColor(ColorStorage, true);
			}
			else
			{
				Console.BackgroundColor = Screensaver.colors[RandomDriver.Next(FallingLineSettings.FallingLineMinimumColorLevel, FallingLineSettings.FallingLineMaximumColorLevel)];
				DebugWriter.WdbgConditional(ref Screensaver.ScreensaverDebug, DebugLevel.I, "Got color ({0})", Console.BackgroundColor);
				ColorStorage = new Color(Console.BackgroundColor);
			}

			// Make the line fall down
			for (int Fall = FallStart, loopTo = FallEnd; Fall <= loopTo; Fall++)
			{
				// Check to see if user decided to resize
				if (CurrentWindowHeight != ConsoleWrapper.WindowHeight | CurrentWindowWidth != ConsoleWrapper.WindowWidth)
					ResizeSyncing = true;
				if (ResizeSyncing)
					break;

				// Print a block and add the covered position to the list so fading down can be done
				TextWriterWhereColor.WriteWherePlain(" ", ColumnLine, Fall, false);
				var PositionTuple = new Tuple<int, int>(ColumnLine, Fall);
				CoveredPositions.Add(PositionTuple);

				// Delay
				ThreadManager.SleepNoBlock(FallingLineSettings.FallingLineDelay, ScreensaverDisplayer.ScreensaverDisplayerThread);
			}

			// Fade the line down. Please note that this requires true-color support in the terminal to work properly.
			for (int StepNum = 0, loopTo1 = FallingLineSettings.FallingLineMaxSteps; StepNum <= loopTo1; StepNum++)
			{
				// Check to see if user decided to resize
				if (CurrentWindowHeight != ConsoleWrapper.WindowHeight | CurrentWindowWidth != ConsoleWrapper.WindowWidth)
					ResizeSyncing = true;
				if (ResizeSyncing)
					break;

				// Set thresholds
				double ThresholdRed = ColorStorage.R / (double)FallingLineSettings.FallingLineMaxSteps;
				double ThresholdGreen = ColorStorage.G / (double)FallingLineSettings.FallingLineMaxSteps;
				double ThresholdBlue = ColorStorage.B / (double)FallingLineSettings.FallingLineMaxSteps;
				DebugWriter.WdbgConditional(ref Screensaver.ScreensaverDebug, DebugLevel.I, "Color threshold (R;G;B: {0})", ThresholdRed, ThresholdGreen, ThresholdBlue);

				// Set color fade steps
				int CurrentColorRedOut = (int)Math.Round(ColorStorage.R - ThresholdRed * StepNum);
				int CurrentColorGreenOut = (int)Math.Round(ColorStorage.G - ThresholdGreen * StepNum);
				int CurrentColorBlueOut = (int)Math.Round(ColorStorage.B - ThresholdBlue * StepNum);
				DebugWriter.WdbgConditional(ref Screensaver.ScreensaverDebug, DebugLevel.I, "Color out (R;G;B: {0};{1};{2})", CurrentColorRedOut, CurrentColorGreenOut, CurrentColorBlueOut);

				// Get the positions and write the block with new color
				var CurrentFadeColor = new Color(CurrentColorRedOut, CurrentColorGreenOut, CurrentColorBlueOut);
				foreach (Tuple<int, int> PositionTuple in CoveredPositions)
				{
					// Check to see if user decided to resize
					if (CurrentWindowHeight != ConsoleWrapper.WindowHeight | CurrentWindowWidth != ConsoleWrapper.WindowWidth)
						ResizeSyncing = true;
					if (ResizeSyncing)
						break;

					// Actually fade the line out
					int PositionLeft = PositionTuple.Item1;
					int PositionTop = PositionTuple.Item2;
					TextWriterWhereColor.WriteWhere(" ", PositionLeft, PositionTop, false, Color.Empty, CurrentFadeColor);
				}

				// Delay
				ThreadManager.SleepNoBlock(FallingLineSettings.FallingLineDelay, ScreensaverDisplayer.ScreensaverDisplayerThread);
			}

			// Reset resize sync
			CoveredPositions.Clear();
			ResizeSyncing = false;
			CurrentWindowWidth = ConsoleWrapper.WindowWidth;
			CurrentWindowHeight = ConsoleWrapper.WindowHeight;
			ThreadManager.SleepNoBlock(FallingLineSettings.FallingLineDelay, ScreensaverDisplayer.ScreensaverDisplayerThread);
		}

	}
}